src/Common/src/CoreWCF/Runtime/TaskHelpers.cs

```diff
@@ -29,120 +30,124 @@ public static async Task AsyncWait<TException>(this Task task)
         // In the EndMethod, you would use ToApmEnd<TResult> to ensure the correct exception handling
         // This will handle throwing exceptions in the correct place and ensure the IAsyncResult contains the provided
         // state object
-        public static Task<TResult> ToApm<TResult>(this Task<TResult> task, AsyncCallback callback, object state)
+        public static IAsyncResult ToApm<T>(this Task<T> task, AsyncCallback callback, object state)
+            => ToApm<T>(new ValueTask<T>(task), callback, state);
+
+        /// <summary>
+        /// Helper method to convert from Task async method to "APM" (IAsyncResult with Begin/End calls)
+        /// </summary>
+        public static IAsyncResult ToApm<T>(this ValueTask<T> task, AsyncCallback callback, object state)
         {
-            // When using APM, the returned IAsyncResult must have the passed in state object stored in AsyncState. This
-            // is so the callback can regain state. If the incoming task already holds the state object, there's no need
-            // to create a TaskCompletionSource to ensure the returned (IAsyncResult)Task has the right state object.
-            // This is a performance optimization for this special case.
-            if (task.AsyncState == state)
+            var result = new AsyncResult<T>(task, callback, state);
+            if (result.CompletedSynchronously)
             {
-                if (callback != null)
-                {
-                    task.ContinueWith((antecedent, obj) =>
-                    {
-                        var callbackObj = obj as AsyncCallback;
-                        callbackObj(antecedent);
-                    }, callback, CancellationToken.None, TaskContinuationOptions.HideScheduler, TaskScheduler.Default);
-                }
-                return task;
+                result.ExecuteCallback();
             }
-
-            // Need to create a TaskCompletionSource so that the returned Task object has the correct AsyncState value.
-            var tcs = new TaskCompletionSource<TResult>(state);
-            var continuationState = Tuple.Create(tcs, callback);
-            task.ContinueWith((antecedent, obj) =>
+            else if (callback != null)
             {
-                var tuple = obj as Tuple<TaskCompletionSource<TResult>, AsyncCallback>;
-                var tcsObj = tuple.Item1;
-                var callbackObj = tuple.Item2;
-                if (antecedent.IsFaulted) tcsObj.TrySetException(antecedent.Exception.InnerException);
-                else if (antecedent.IsCanceled) tcsObj.TrySetCanceled();
-                else tcsObj.TrySetResult(antecedent.Result);
-
-                if (callbackObj != null)
+                task.AsTask().ContinueWith((res, asyncResult) =>
```

Comment:
do you prefer to use "_" (discard) for unused parameter?

Reply:
I do prefer the new discard feature. It's quite a nice addition to the language.